cdap-security/src/test/java/co/cask/cdap/security/authorization/AuthorizerTest.java

```diff
@@ -0,0 +1,111 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.security.authorization;
+
+import co.cask.cdap.common.UnauthorizedException;
+import co.cask.cdap.proto.id.DatasetId;
+import co.cask.cdap.proto.id.EntityId;
+import co.cask.cdap.proto.id.Ids;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.security.Action;
+import co.cask.cdap.proto.security.Principal;
+import com.google.common.collect.ImmutableSet;
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ * Tests for {@link Authorizer}.
+ */
+public abstract class AuthorizerTest {
+
+  private final NamespaceId namespace = Ids.namespace("foo");
+  private final Principal user = new Principal("alice", Principal.PrincipalType.USER);
+
+  protected abstract Authorizer get();
+
+  @Test
+  public void testSimple() throws UnauthorizedException {
+    Authorizer authorizer = get();
+
+    verifyAuthFailure(namespace, user, Action.READ);
+
+    authorizer.grant(namespace, user, ImmutableSet.of(Action.READ));
+    authorizer.enforce(namespace, user, Action.READ);
+
+    authorizer.revoke(namespace, user, ImmutableSet.of(Action.READ));
+    verifyAuthFailure(namespace, user, Action.READ);
+  }
+
+  @Test
+  public void testWildcard() throws UnauthorizedException {
+    Authorizer authorizer = get();
+
+    verifyAuthFailure(namespace, user, Action.READ);
+
+    authorizer.grant(namespace, user);
+    authorizer.enforce(namespace, user, Action.READ);
+    authorizer.enforce(namespace, user, Action.WRITE);
+
+    authorizer.revoke(namespace, user);
+    verifyAuthFailure(namespace, user, Action.READ);
+  }
+
+  @Test
+  public void testAll() throws UnauthorizedException {
+    Authorizer authorizer = get();
+
+    verifyAuthFailure(namespace, user, Action.READ);
+
+    authorizer.grant(namespace, user, ImmutableSet.of(Action.ALL));
+    authorizer.enforce(namespace, user, Action.READ);
+    authorizer.enforce(namespace, user, Action.WRITE);
+
+    authorizer.revoke(namespace, user);
+    verifyAuthFailure(namespace, user, Action.READ);
+
+    Principal role = new Principal("admins", Principal.PrincipalType.ROLE);
+    authorizer.grant(namespace, user, ImmutableSet.of(Action.READ));
+    authorizer.grant(namespace, role, ImmutableSet.of(Action.ALL));
+    authorizer.revoke(namespace);
+    verifyAuthFailure(namespace, user, Action.READ);
+    verifyAuthFailure(namespace, role, Action.ALL);
+  }
+
+  @Test
+  public void testHierarchy() throws UnauthorizedException {
+    Authorizer authorizer = get();
+
+    DatasetId dataset = namespace.dataset("bar");
+
+    verifyAuthFailure(namespace, user, Action.READ);
+
+    authorizer.grant(namespace, user, ImmutableSet.of(Action.READ));
+    authorizer.enforce(dataset, user, Action.READ);
```

Comment:
should we also have other way here granting write on dataset does not grant write on namesapce